mlem/contrib/kubernetes/base.py

```diff
@@ -0,0 +1,281 @@
+        node_index = find_index(node_list, node_name)
+        address_dict = node_list[node_index].status.addresses
+        for each_address in address_dict:
+            if each_address.type == "ExternalDNS":
+                host = each_address.address
+        return host, port
+
+
+class LoadBalancerService(ServiceType):
+
+    type: ClassVar = "k8s_loadbalancer_service"
+
+    def get_host_and_port(self, service, namespace="mlem"):
+        host, port = None, None
+        port = service.items[0].spec.ports[0].port
+        ingress = service.items[0].status.load_balancer.ingress[0]
+        host = ingress.hostname or ingress.ip
+        return host, port
+
+
+class K8sDeploymentState(DeployState):
+    """DeployState implementation for Kubernetes deployments"""
+
+    type: ClassVar = "kubernetes"
+
+    image: Optional[DockerImage] = None
+    """Docker Image being used for Deployment"""
+    deployment_name: Optional[str] = None
+    """Name of Deployment"""
+
+
+class K8sDeployment(MlemDeployment, K8sYamlBuildArgs):
+    """MlemDeployment implementation for Kubernetes deployments"""
+
+    type: ClassVar = "kubernetes"
+    state_type: ClassVar = K8sDeploymentState
+    """type of state for Kubernetes deployments"""
+
+    server: Optional[Server] = None
+    """type of Server to use, with options such as FastAPI, RabbitMQ etc."""
+    registry: Optional[DockerRegistry] = DockerRegistry()
+    """docker registry"""
+    daemon: Optional[DockerDaemon] = DockerDaemon(host="")
+    """docker daemon"""
+    kube_config_file_path: Optional[str] = None
+    """path for kube config file of the cluster"""
+    templates_dir: List[str] = []
+    """list of dirs where templates reside"""
+
+    def load_kube_config(self):
+        config.load_kube_config(
+            config_file=self.kube_config_file_path
+            or os.getenv("KUBECONFIG", default="~/.kube/config")
+        )
+
+    def _get_client(self, state: K8sDeploymentState) -> Client:
+        host, port = None, None
+        self.load_kube_config()
+        service = client.CoreV1Api().list_namespaced_service(self.namespace)
+        try:
+            if self.service_type == ServiceTypeEnum.node_port:
+                host, port = NodePortService().get_host_and_port(
+                    service, self.namespace
+                )
+            elif self.service_type == ServiceTypeEnum.load_balancer:
+                host, port = LoadBalancerService().get_host_and_port(service)
+            elif self.service_type == ServiceTypeEnum.cluster_ip:
+                raise TypeError(
+                    "Cannot expose service of type ClusterIP outside the Kubernetes Cluster"
+                )
+            else:
+                raise ValueError(
+                    f"service_type supplied is {self.service_type}, valid values are {[e.value for e in ServiceTypeEnum]} only"
+                )
+        except (IndexError, ValueError, TypeError) as e:
```

Comment:
Those checks should be inside corresponding `get_host_and_port` implementations, not common. Those index errors can mean different things for different implementations and even different things for one implementation

Reply:
Index errors and attribute errors are common really. They don't mean different things. Type Error and Value error is what I added because I was explicitly raising `TypeError` for ClusterIP service and ValueError when the service type was something different than what is allowed in `Enum`